synapse/handlers/room_member.py

```diff
@@ -823,8 +841,14 @@ async def update_membership_locked(
 
         latest_event_ids = await self.store.get_prev_events_for_room(room_id)
 
+        # There are 2 reasons that mandate to have the full state:
+        # - to calculate `is_host_in_room`, however we now approximate that we consider the server
+        # in the room if we are still in the middle of a fast join.
+        # - to calculate if another user trying to join is allowed to. In this case we now
+        # make another `make/send_join` to a resident server to validate that. We could just
+        # accept it but it would mean that we would potentially leak the history to a banned user.
```

Comment:
Not fully sure what you're getting at here. Are you trying to enumerate the ways in which we use `state_before_join` and argue that it's okay that if this is only partial state?

Reply:
Many thanks, this is much appreciated.